### PR TITLE
[Core] Clean up builder and solvers

### PR DESCRIPTION
**Pull request by loumalouomega**
Basically cleans the base class for builder and solver and the elimination builder and solver (and it add the Doxygen documentation)

This change is aesthetic, and doesn't change the behavior of the B&S

This was part of #3236

### CODE REVIEW

**Review by RiccardoRossi — COMMENTED** [review]
@loumalouomega , i block this until other more important changes in the BuilderAndSolvers go ahead
(for example the one of adding MPCs to the Elimination). As you are saying this would cause conflicts and would render the review of the other PRs more diffcult

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h]: this change is NOT aesthetic!

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
i am blocking since you are pushing as "aesthetic" one change that is not. AND IN A CLASS OF MAJOR IMPORTANCE!

My personal feeling is that i'd rather avoid changes for aesthetic reasons in important classes. in any case i'll ask for @KratosMultiphysics/implementation-committee 's opinion on this. (@loumalouomega please recall that to issue a reccomendation you should reach a consensus in the committee)

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h]: This was already blocked, I just saw that it wasn't compiling due to a merge conflict. I only solved that, and you already said that are other things first

**Comment by loumalouomega** [comment]
I need to check after merging with the master that the B&S is still working

**Comment by loumalouomega** [comment]
@RiccardoRossi after last changes this is just a clean up

**Comment by pablobecker** [comment]
Is this now 100% aesthetic, or are other changes as Riccardo found?
 It's more than 1K lines of codes,  so it's easy to miss something

**Comment by loumalouomega** [comment]
Most of the changes are Doxygen documentation, if you select the no WhiteSpaces the changes reduce to 800lines aprox

**Comment by GuillermoCasas** [comment]
From the @KratosMultiphysics/implementation-committee, there is no problem with this PR, since we have checked that it consists purely in aesthetic changes (to the better). We pass it on to the @KratosMultiphysics/technical-committee for the final evaluation.

**Comment by RiccardoRossi** [comment]
i think this is superseded by the latest changes of the builder and sovlers.

@loumalouomega can we close this?

**Comment by loumalouomega** [comment]
> i think this is superseded by the latest changes of the builder and sovlers.
> 
> @loumalouomega can we close this?

Not at all, this is a clean up. No real change, can you take a look?

**Comment by loumalouomega** [comment]
> > i think this is superseded by the latest changes of the builder and sovlers.
> > @loumalouomega can we close this?
> 
> Not at all, this is a clean up. No real change, can you take a look?

Look at teh consensus of the @KratosMultiphysics/implementation-committee

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: there is also 4 - print to Matrix Market

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: do not define this. use TSystemMAtrixType instead

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:1237]: ...this has no ward? i mean, it is not threadsafe...

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: no ward? (not threadsafe)

**Review by RiccardoRossi — COMMENTED** [review]
ok, i took a look. I am close to accepting, nevertheless take a look at my comments please

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:1237]: As you can see this was already like this....and maybe the reason it was failing the RVE test...

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: Gasp, probably you found just for coincidence the reason of the fail on the RVE test

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: OK

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: const

**Review by philbucher — COMMENTED** [review]
minor comment, otehrwise this looks good to me

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:1237]: This assemble is not in parallel...

Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:1237]: Does not solve the RVE test :(

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:766]: I believe you miss adding the dofs which are introduced by the MPCs...

Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:766]: This method is override on the MPC version of the B&S

Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:766]: i thought we unified this in one class...

ok

Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:766]: Not yet, and is not the target of this PR

**Comment by loumalouomega** [comment]
Ping @RiccardoRossi

**Comment by loumalouomega** [comment]
Ping

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: See next line

**Comment by loumalouomega** [comment]
I fixed a conflict, I remember to @KratosMultiphysics/technical-committee that this PR now just documents and cleans ups the classes and there is already a consensus from @KratosMultiphysics/implementation-committee

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: @loumalouomega to me it seems like this is also refactored
The rest seems fine for me, but I don't feel comfortable approving due to this

Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: I don't understand, I must remove this method?

Comment (philbucher) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: I would do this change in a separate PR, makes it easier to review

Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: Ok, removed and now is in concordance with the master

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (philbucher) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: is this solved?

**Comment by loumalouomega** [comment]
Yes

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: this is refactored

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (philbucher) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: ping

**Review by philbucher — COMMENTED** [review]
to me looks ok, I haven't found an obvious error

this is really tough to review, many changes. Please make smaller PRs for such things.

up to you @RiccardoRossi

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: More than refactored is that the code has been change in the master recently; i will update

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:196]: why should we do this?

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:336]: I think this is not correct since AssembleLHS doesn't have the atomic guards needed

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee tried to review this PR and decide to block it. As this happens in many PR we also decided to emphasis the problem of such PRs where:

1. Doing a lot of "Cosmetic" changes in a very core file.
2. Mixing the "Cosmetic" changes with real changes. Even changing the whitespaces should go in a separate PR
3. The "Cosmetic change" cannot include any change in the behavior of the code. So in case that it does change the behavior, the PR should be blocked
4. Please consider that using auto instead of type or changing constantness IS NOT cosmetic and actually source of very difficult to detect errors. (like missing the reference and copy the object, etc...)

The rational behind is the fact that this type of PRs are very prone to make pass some unwanted errors to the code, regardless the effort of the writer or the reviewer.

@loumalouomega would you please close this PR and open few of the ones with specific formatting changes and then one with the behavior changes?

**Comment by loumalouomega** [comment]
Ping @RiccardoRossi

**Comment by philbucher** [comment]
> @KratosMultiphysics/technical-committee tried to review this PR and decide to block it. As this happens in many PR we also decided to emphasis the problem of such PRs where:
> 
>     1. Doing a lot of "Cosmetic" changes in a very core file.
> 
>     2. Mixing the "Cosmetic" changes with real changes. Even changing the whitespaces should go in a separate PR
> 
>     3. The "Cosmetic change" cannot include any change in the behavior of the code. So in case that it does change the behavior, the PR should be blocked
> 
>     4. Please consider that using auto instead of type or changing constantness IS NOT cosmetic and actually source of very difficult to detect errors. (like missing the reference and copy the object, etc...)
> 
> 
> The rational behind is the fact that this type of PRs are very prone to make pass some unwanted errors to the code, regardless the effort of the writer or the reviewer.
> 
> @loumalouomega would you please close this PR and open few of the ones with specific formatting changes and then one with the behavior changes?

@loumalouomega

**Comment by loumalouomega** [comment]
> 
> 
> > @KratosMultiphysics/technical-committee tried to review this PR and decide to block it. As this happens in many PR we also decided to emphasis the problem of such PRs where:
> > ```
> > 1. Doing a lot of "Cosmetic" changes in a very core file.
> > 
> > 2. Mixing the "Cosmetic" changes with real changes. Even changing the whitespaces should go in a separate PR
> > 
> > 3. The "Cosmetic change" cannot include any change in the behavior of the code. So in case that it does change the behavior, the PR should be blocked
> > 
> > 4. Please consider that using auto instead of type or changing constantness IS NOT cosmetic and actually source of very difficult to detect errors. (like missing the reference and copy the object, etc...)
> > ```
> > 
> > 
> > The rational behind is the fact that this type of PRs are very prone to make pass some unwanted errors to the code, regardless the effort of the writer or the reviewer.
> > @loumalouomega would you please close this PR and open few of the ones with specific formatting changes and then one with the behavior changes?
> 
> @loumalouomega

OK, I did not saw that

**Comment by loumalouomega** [comment]
> 
> 
> > > @KratosMultiphysics/technical-committee tried to review this PR and decide to block it. As this happens in many PR we also decided to emphasis the problem of such PRs where:
> > > ```
> > > 1. Doing a lot of "Cosmetic" changes in a very core file.
> > > 
> > > 2. Mixing the "Cosmetic" changes with real changes. Even changing the whitespaces should go in a separate PR
> > > 
> > > 3. The "Cosmetic change" cannot include any change in the behavior of the code. So in case that it does change the behavior, the PR should be blocked
> > > 
> > > 4. Please consider that using auto instead of type or changing constantness IS NOT cosmetic and actually source of very difficult to detect errors. (like missing the reference and copy the object, etc...)
> > > ```
> > > 
> > > 
> > > The rational behind is the fact that this type of PRs are very prone to make pass some unwanted errors to the code, regardless the effort of the writer or the reviewer.
> > > @loumalouomega would you please close this PR and open few of the ones with specific formatting changes and then one with the behavior changes?
> > 
> > 
> > @loumalouomega
> 
> OK, I did not saw that

I think this is ready

**Comment by philbucher** [comment]
this has still a lot of changes, I cannot tell if all of them are cosmetic
I strongly suggest to close this PR and make smaller ones
I lost the overview and am scared of it :/

**Comment by loumalouomega** [comment]
> 
> 
> this has still a lot of changes, I cannot tell if all of them are cosmetic
> I strongly suggest to close this PR and make smaller ones
> I lost the overview and am scared of it :/

Most of teh changes are cosmetic, I can separete more code but there is a point where one cosmetic change propagates (like adapting variable names to the style guide)

**Comment by pooyan-dadvand** [comment]
I agree with @philbucher that there are still too many changes (~1000 lines changed!). IMO:
1. The changes in documentation should go in one PR
2. The ones of the variable name changes should be only renaming and nothing more in the PR.
3. The one of repositioning the code or brackets are the most annoying one and I would avoid them for now.

**Comment by RiccardoRossi** [comment]
this is a large PR and also intrusive (and risky because of where it is). @KratosMultiphysics/technical-committee proposes closing. We should consider the changes as @pooyan-dadvand commented in his previous post

**Comment by loumalouomega** [comment]
> 
> 
> this is a large PR and also intrusive (and risky because of where it is). @KratosMultiphysics/technical-committee proposes closing. We should consider the changes as @pooyan-dadvand commented in his previous post

Please, do not remove the branch until I copied what I need

**Comment by loumalouomega** [comment]
After #6591 this PR has been simplified, I can still split it in smaller parts

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (pablobecker) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:329]: so previously this was in serial and not using the active flag?
shouldnt you use the locks here too?

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (pablobecker) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:414]: also here

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (pablobecker) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:820]: why?

**Review by RiccardoRossi — COMMENTED** [review]
some comments

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:820]: To simplify the code

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:196]: At the end is the same, and is more clear

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:336]: I will check, but i would say it has

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:329]: I will check

**Comment by adityaghantasala** [comment]
On behalf of @KratosMultiphysics/implementation-committee  

Please keep this PR completely about clean up (only aesthetic changes) so remove the pragmas. Make a separate a PR with these changes and the corresponding developers can have a look a them in deep. Please have a look at @pablobecker 's comment about the ACTIVE elements, this also should be in the new PR.

**Comment by pablobecker** [comment]
ping @loumalouomega

**Pull request closed without merging**
